### PR TITLE
Introduce cache entry/table utilities

### PR DESCRIPTION
**Pull request by SaitTalhaNisanci**
In the codebase, when we need to learn the type of a table we use different methods, one of the common ones is to check the partition method, for example we say that a table is a reference table if its partition method is `DISTRIBUTE_BY_NONE`. This creates a mapping in the code, a developer doesn't need to know this mapping. In the codebase what we care about is almost always is the following two:
- What is the type of citus cache entry, is it reference table, hash distributed table etc?
- What is the type of table with given `oid`, is is reference table, hash distributed table etc?

In order to accomplish this, some utility functions are created. This also makes it easier to add new table types to the codebase. For example when we add citus-local tables, they will also have `DISTRIBUTE_BY_NONE` as their replication model hence all the lines that had this check for reference tables will possibly fail. There are some parts where we care about if a table is `reference table` and some parts where we care it is a `distributed table` (hash, range, append) or if it is a table without a partition key.

There are still some places with `DISTRIBUTE_BY_NONE`  etc checks, we can possibly do the refactoring for them separately since for them we pass the partition method as a parameter to function(so relation id or cache entry doesn't exist).

The new APIs:

```c
IsCitusTableType(relationId, CitusTableType);  // this will be used when we don't have cache entry but relation id, this does an extra lookup from metadata cache
IsCitusTableTypeCacheEntry(cacheEntry, CitusTableType); // this will be used when we have cache entry
```

Where a table types can be:

```c
typedef enum
{
	HASH_DISTRIBUTED,
	APPEND_DISTRIBUTED,
	RANGE_DISTRIBUTED,
	DISTRIBUTED_TABLE, /* hash, range or append distributed table */
	REFERENCE_TABLE,
	CITUS_LOCAL_TABLE,
	CITUS_TABLE_WITH_NO_DIST_KEY /* table without a dist key such as reference table */
} CitusTableType;
```

### CODE REVIEW

File: src/backend/distributed/metadata/metadata_cache.c [hunk 1]
```diff
@@ -296,6 +296,191 @@ EnsureModificationsCanRun(void)
...
+	return IsNonDistributedTableCacheEntry(tableEntry);
+}
+
+
+/*
+ * IsReferenceTable returns whether the given relation ID identifies a reference
+ * table.
+ */
+bool
+IsReferenceTable(Oid relationId)
```
Comment (onderkalaci) [inline src/backend/distributed/metadata/metadata_cache.c:320]: One point is that using `IsReferenceTable (relationId)` is tempting over `IsReferenceTableCacheEntry ()` even if we already have the `tableCacheEntry` at hand. The callers are likely to forget about `IsReferenceTableCacheEntry()`

 And, this may have performance implications in certain code-paths.

What if we change the  functions to something like and get rid of `IsReferenceTableCacheEntry `?:

```C
IsReferenceTable(Oid  relationId, CitusTableCacheEntry *optionalTableEntry)
{
     /* if the optionalTableEntry  entry is not NULL,  we're good. 
       *  else we get the cache entry and do the same*/
}
```

This may not be the best API, just a quick suggestion. Do you have any ideas on  this?

(One idea could be to verify that my concern  about performance is not relevant to the  discussion at all)

File: src/backend/distributed/metadata/metadata_cache.c [hunk 2]
```diff
@@ -296,6 +296,191 @@ EnsureModificationsCanRun(void)
...
 
 
+/*
+ * IsNonDistributedTable returns whether the given relation ID identifies a
+ * non-distributed table such as reference table.
+ */
+bool
+IsNonDistributedTable(Oid relationId)
+{
+	if (!IsCitusTable(relationId))
```
Comment (onderkalaci) [inline src/backend/distributed/metadata/metadata_cache.c:306]: `IsCitusTable` and `IsNonDistributedTableCacheEntry ` may rely on different `CitusTableCacheEntry `. What happens if `GetCitusTableCacheEntry ` invalidates the cache entry in between? Would that be OK for all the functions/code-paths?

**Review by onderkalaci — COMMENTED** [review]
I'm sharing my initial feedback/questions. Please have a look and we can discuss over teams as well

File: src/backend/distributed/metadata/metadata_cache.c [hunk 3]
```diff
@@ -296,6 +296,191 @@ EnsureModificationsCanRun(void)
...
+
+
+/*
+ * IsDistributedTable returns whether the given relation ID identifies a distributed
+ * table. A distributed table can be of hash, range or append type.
+ */
+bool
+IsDistributedTable(Oid relationId)
+{
+	if (!IsCitusTable(relationId))
```
[the remark this answers is not in the document]
Comment (SaitTalhaNisanci) [reply src/backend/distributed/metadata/metadata_cache.c:338]: We can return the cache entry from a variation of IsCitusTable to reduce the call.

I thought that the actual cost of calling LookupCitusTableCacheEntry is going to be low because we will do a simple lookup in most of the cases from a hash table. However, I havent don't any benchmark, do you have any data on this?

[the remark this answers is not in the document]
Comment (onderkalaci) [reply src/backend/distributed/metadata/metadata_cache.c:338]: I don't have any data, that's why I'm trying to be cautious.

File: src/include/distributed/metadata_cache.h [hunk 4]
```diff
@@ -116,7 +116,19 @@ typedef struct DistObjectCacheEntry
 	int colocationId;
 } DistObjectCacheEntry;
 
-
+extern bool IsNonDistributedTable(Oid relationId);
```
Comment (onderkalaci) [inline src/include/distributed/metadata_cache.h:119]: please add comment

File: src/include/distributed/metadata_cache.h [hunk 5]
```diff
@@ -116,7 +116,19 @@ typedef struct DistObjectCacheEntry
...
+extern bool IsReferenceTable(Oid relationId);
+extern bool IsDistributedTable(Oid relationId);
+extern bool IsAppendDistributedTable(Oid relationId);
+extern bool IsHashDistributedTable(Oid relationId);
+extern bool IsRangeDistributedTable(Oid relationId);
```
Comment (onderkalaci) [inline src/include/distributed/metadata_cache.h:124]: Maybe `IsCitusTableType(oid, enum)` to avoid the word `Distributed`

```
HASH_DISTRIBUTED, APPEND_DISTRIBUTED, RANGE_DISTRIBUTED,
REFERENCE_TABLE, CITUS_LOCAL_TABLE, CITUS_TABLE_WITH_NO_DIST_KEY
```

File: src/backend/distributed/operations/stage_protocol.c [hunk 6]
```diff
@@ -136,15 +136,14 @@ master_create_empty_shard(PG_FUNCTION_ARGS)
...
-	if (partitionMethod == DISTRIBUTE_BY_HASH)
+	if (IsCitusTableType(relationId, HASH_DISTRIBUTED))
 	{
 		ereport(ERROR, (errmsg("relation \"%s\" is a hash partitioned table",
 							   relationName),
 						errdetail("We currently don't support creating shards "
 								  "on hash-partitioned tables")));
 	}
-	else if (partitionMethod == DISTRIBUTE_BY_NONE)
+	else if (IsCitusTableType(relationId, REFERENCE_TABLE))
```
Comment (onderkalaci) [inline src/backend/distributed/operations/stage_protocol.c:146]: I guess we should  use `NO_DIST_KEY` as `master_create_empty_shard ` should error out for local tables.

(I don't care much about the error message, we can live with that)

File: src/backend/distributed/operations/stage_protocol.c [hunk 7]
```diff
@@ -253,8 +252,8 @@ master_append_table_to_shard(PG_FUNCTION_ARGS)
 						errdetail("The underlying shard is not a regular table")));
 	}
 
-	char partitionMethod = PartitionMethod(relationId);
-	if (partitionMethod == DISTRIBUTE_BY_HASH || partitionMethod == DISTRIBUTE_BY_NONE)
+	if (IsCitusTableType(relationId, HASH_DISTRIBUTED) || IsCitusTableType(relationId,
```
Comment (onderkalaci) [inline src/backend/distributed/operations/stage_protocol.c:255]: Same here, we don't need to support local tables `master_append_table_to_shard `, so should better be `NO_DIST_KEY`

File: src/backend/distributed/operations/stage_protocol.c [hunk 8]
```diff
@@ -726,7 +725,7 @@ WorkerCreateShardCommandList(Oid relationId, int shardIndex, uint64 shardId,
 		{
 			referencedShardId = shardId;
 		}
-		else if (PartitionMethod(referencedRelationId) == DISTRIBUTE_BY_NONE)
+		else if (IsCitusTableType(referencedRelationId, REFERENCE_TABLE))
```
Comment (onderkalaci) [inline src/backend/distributed/operations/stage_protocol.c:728]: these are tricky ones :) In theory, we could use `DIST_NONE` instead of reference table, but this code path is not triggered by local tables (we don't support local - distributed foreign keys yet). But, I'm good with the currentt selection

File: src/backend/distributed/planner/distributed_planner.c [hunk 9]
```diff
@@ -306,8 +306,7 @@ ExtractReferenceTableRTEList(List *rteList)
 		}
 
 		Oid relationOid = rte->relid;
-		if (IsCitusTable(relationOid) && PartitionMethod(relationOid) ==
-			DISTRIBUTE_BY_NONE)
+		if (IsCitusTableType(relationOid, REFERENCE_TABLE))
```
Comment (onderkalaci) [inline src/backend/distributed/planner/distributed_planner.c:309]: we seem to not use this function, feel free to remove

File: src/backend/distributed/planner/multi_join_order.c [hunk 10]
```diff
@@ -823,12 +823,13 @@ ReferenceJoin(JoinOrderNode *currentJoinNode, TableEntry *candidateTable,
...
 	}
 
-	char candidatePartitionMethod = PartitionMethod(candidateTable->relationId);
-	char leftPartitionMethod = PartitionMethod(currentJoinNode->tableEntry->relationId);
-
 	if (!IsSupportedReferenceJoin(joinType,
-								  leftPartitionMethod == DISTRIBUTE_BY_NONE,
-								  candidatePartitionMethod == DISTRIBUTE_BY_NONE))
+								  IsCitusTableType(
+									  currentJoinNode->tableEntry->relationId,
```
Comment (onderkalaci) [inline src/backend/distributed/planner/multi_join_order.c:828]: could use intermediate variables, pretty hard to follow
```
bool leftIsReferenceTable = ....
bool rightIsReference = ...
```
as these are the parameters that `IsSupportedReferenceJoin ` gets

File: src/backend/distributed/planner/multi_join_order.c [hunk 11]
```diff
@@ -874,12 +875,12 @@ static JoinOrderNode *
 CartesianProductReferenceJoin(JoinOrderNode *currentJoinNode, TableEntry *candidateTable,
 							  List *applicableJoinClauses, JoinType joinType)
 {
-	char candidatePartitionMethod = PartitionMethod(candidateTable->relationId);
-	char leftPartitionMethod = PartitionMethod(currentJoinNode->tableEntry->relationId);
-
 	if (!IsSupportedReferenceJoin(joinType,
-								  leftPartitionMethod == DISTRIBUTE_BY_NONE,
-								  candidatePartitionMethod == DISTRIBUTE_BY_NONE))
+								  IsCitusTableType(
```
Comment (onderkalaci) [inline src/backend/distributed/planner/multi_join_order.c:879]: again, could use variables

File: src/backend/distributed/planner/multi_logical_planner.c [hunk 12]
```diff
@@ -227,10 +227,10 @@ TargetListOnPartitionColumn(Query *query, List *targetEntryList)
 		FindReferencedTableColumn(targetExpression, NIL, query, &relationId, &column);
 
 		/*
-		 * If the expression belongs to a reference table continue searching for
+		 * If the expression belongs to a non-distributed table continue searching for
 		 * other partition keys.
 		 */
-		if (IsCitusTable(relationId) && PartitionMethod(relationId) == DISTRIBUTE_BY_NONE)
+		if (IsCitusTableType(relationId, CITUS_TABLE_WITH_NO_DIST_KEY))
```
Comment (onderkalaci) [inline src/backend/distributed/planner/multi_logical_planner.c:233]: I guess we cannot get here with local tables, but seems fine to keep as-is, I cannot see any harm

File: src/backend/distributed/planner/multi_physical_planner.c [hunk 13]
```diff
@@ -2141,7 +2141,7 @@ QueryPushdownSqlTaskList(Query *query, uint64 jobId,
 		ListCell *shardIntervalCell = NULL;
 
 		CitusTableCacheEntry *cacheEntry = GetCitusTableCacheEntry(relationId);
-		if (cacheEntry->partitionMethod == DISTRIBUTE_BY_NONE)
+		if (IsCitusTableTypeCacheEntry(cacheEntry, CITUS_TABLE_WITH_NO_DIST_KEY))
```
Comment (onderkalaci) [inline src/backend/distributed/planner/multi_physical_planner.c:2144]: we cannot have local tables here, still no harm..

File: src/backend/distributed/planner/multi_physical_planner.c [hunk 14]
```diff
@@ -2537,13 +2536,13 @@ CoPartitionedTables(Oid firstRelationId, Oid secondRelationId)
 	FmgrInfo *comparisonFunction = firstTableCache->shardIntervalCompareFunction;
 
 	/* reference tables are always & only copartitioned with reference tables */
-	if (firstTableCache->partitionMethod == DISTRIBUTE_BY_NONE &&
-		secondTableCache->partitionMethod == DISTRIBUTE_BY_NONE)
+	if (IsCitusTableTypeCacheEntry(firstTableCache, REFERENCE_TABLE) &&
```
Comment (onderkalaci) [inline src/backend/distributed/planner/multi_physical_planner.c:2539]: to be compatible with other planner functions, let's use NONE_DIST_KEY in this function?

File: src/backend/distributed/planner/multi_router_planner.c [hunk 15]
```diff
@@ -2702,14 +2699,13 @@ BuildRoutesForInsert(Query *query, DeferredErrorMessage **planningError)
...
 	CitusTableCacheEntry *cacheEntry = GetCitusTableCacheEntry(distributedTableId);
-	char partitionMethod = cacheEntry->partitionMethod;
 	List *modifyRouteList = NIL;
 	ListCell *insertValuesCell = NULL;
 
 	Assert(query->commandType == CMD_INSERT);
 
 	/* reference tables can only have one shard */
-	if (partitionMethod == DISTRIBUTE_BY_NONE)
+	if (IsCitusTableTypeCacheEntry(cacheEntry, REFERENCE_TABLE))
```
Comment (onderkalaci) [inline src/backend/distributed/planner/multi_router_planner.c:2708]: Shouldn't this be `CITUS_TABLE_WITH_NO_DIST_KEY `? We'll be using this code-path for planning queries

File: src/backend/distributed/transaction/relation_access_tracking.c [hunk 16]
```diff
@@ -173,7 +173,7 @@ RecordRelationAccessIfReferenceTable(Oid relationId, ShardPlacementAccessType ac
 	 * recursively calling RecordRelationAccessBase(), so becareful about
 	 * removing this check.
 	 */
-	if (PartitionMethod(relationId) != DISTRIBUTE_BY_NONE)
+	if (!IsCitusTableType(relationId, REFERENCE_TABLE))
```
Comment (onderkalaci) [inline src/backend/distributed/transaction/relation_access_tracking.c:176]: I'm not sure about this, though we can finalize this part with @onurctirtir tomorrow on this PR, let's merge this as-is for now

File: src/backend/distributed/utils/colocation_utils.c [hunk 17]
```diff
@@ -431,11 +431,11 @@ ShardsIntervalsEqual(ShardInterval *leftShardInterval, ShardInterval *rightShard
 		return false;
 	}
 
-	if (leftIntervalPartitionMethod == DISTRIBUTE_BY_HASH)
+	if (IsCitusTableType(leftShardInterval->relationId, HASH_DISTRIBUTED))
 	{
 		return HashPartitionedShardIntervalsEqual(leftShardInterval, rightShardInterval);
 	}
-	else if (leftIntervalPartitionMethod == DISTRIBUTE_BY_NONE)
+	else if (IsCitusTableType(leftShardInterval->relationId, REFERENCE_TABLE))
```
Comment (onderkalaci) [inline src/backend/distributed/utils/colocation_utils.c:438]: I guess we're good to use `NO_DIST_KEY` here, because they also have single shard

File: src/backend/distributed/utils/shardinterval_utils.c [hunk 18]
```diff
@@ -240,21 +240,22 @@ ShardIndex(ShardInterval *shardInterval)
...
+			cacheEntry, REFERENCE_TABLE))
 	{
 		ereport(ERROR, (errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
 						errmsg("finding index of a given shard is only supported for "
 							   "hash distributed and reference tables")));
 	}
 
 	/* short-circuit for reference tables */
-	if (partitionMethod == DISTRIBUTE_BY_NONE)
+	if (IsCitusTableTypeCacheEntry(cacheEntry, REFERENCE_TABLE))
```
Comment (onderkalaci) [inline src/backend/distributed/utils/shardinterval_utils.c:258]: I guess this could use `NO_DIST_KEY` (same for above)

File: src/include/distributed/metadata_cache.h [hunk 19]
```diff
@@ -116,6 +116,20 @@ typedef struct DistObjectCacheEntry
 	int colocationId;
 } DistObjectCacheEntry;
 
+typedef enum
+{
+	HASH_DISTRIBUTED,
+	APPEND_DISTRIBUTED,
+	RANGE_DISTRIBUTED,
+	DISTRIBUTED_TABLE, /* hash, range or append distributed table */
```
Comment (onderkalaci) [inline src/include/distributed/metadata_cache.h:124]: I guess the style can be improved a little

```C
typedef enum
{
	HASH_DISTRIBUTED,
	APPEND_DISTRIBUTED,
	RANGE_DISTRIBUTED,

        /* hash, range or append distributed table */
	DISTRIBUTED_TABLE, 

	REFERENCE_TABLE,
	CITUS_LOCAL_TABLE,

        /* table without a dist key such as reference table and citus local table */
	CITUS_TABLE_WITH_NO_DIST_KEY 
} CitusTableType;
```

**Review by onderkalaci — APPROVED** [review]
This is a pretty big PR, I went over every line of change carefully, as it might have bad side-effects.

Overall, the PR is simplifying the code and making citus-local table integration much smoother.

Please address my review, I'm giving conditional approval. If you do not agree with any of the review, please wait before the merge and we can discuss it.

Thanks!

File: src/backend/distributed/commands/truncate.c [hunk 20]
```diff
@@ -317,8 +316,7 @@ ExecuteTruncateStmtSequentialIfNecessary(TruncateStmt *command)
 	{
 		Oid relationId = RangeVarGetRelid(rangeVar, NoLock, failOK);
 
-		if (IsCitusTable(relationId) &&
-			PartitionMethod(relationId) == DISTRIBUTE_BY_NONE &&
+		if (IsCitusTableType(relationId, REFERENCE_TABLE) &&
```
[the remark this answers is not in the document]
Comment (SaitTalhaNisanci) [reply src/backend/distributed/commands/truncate.c:319]: `IsCitusTableType` will return false if it is not a citus table so it seems that the behavior is the same as we don't error out in `IsCitusTable`.
